src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -0,0 +1,335 @@
+        /// <param name="getActivityDataRequestUsingContext">The callback which will get called every time the ActivitySource try to creates a new Activity object using ActivityContext.</param>
+        /// <param name="getActivityDataRequestUsingParentId">The callback which will get called every time the ActivitySource try to creates a new Activity object using ParentId.</param>
+        /// <param name="onActivityStarted">The callback which get called everytime the activity get started.</param>
+        /// <param name="onActivityStopped">The callback which get called everytime the activity get stopped.</param>
+        /// <returns>The disposable object represent the listener. When disposing the listener, it will stop the Activity event notification.</returns>
+        public static IDisposable AddActivityListener(
+                                System.Func<System.Diagnostics.ActivitySource, bool> listenToSource,
+                                System.Func<System.Diagnostics.ActivitySource, string, System.Diagnostics.ActivityKind, System.Diagnostics.ActivityContext, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, string?>>?, System.Collections.Generic.IEnumerable<System.Diagnostics.ActivityLink>?, System.Diagnostics.ActivityDataRequest> getActivityDataRequestUsingContext,
+                                System.Func<System.Diagnostics.ActivitySource, string, System.Diagnostics.ActivityKind, string, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, string?>>?, System.Collections.Generic.IEnumerable<System.Diagnostics.ActivityLink>?, System.Diagnostics.ActivityDataRequest> getActivityDataRequestUsingParentId,
+                                Action<Activity> onActivityStarted,
+                                Action<Activity> onActivityStopped)
+        {
+            ActivityListener listener = new ActivityListener(listenToSource, getActivityDataRequestUsingContext, getActivityDataRequestUsingParentId, onActivityStarted, onActivityStopped);
+            s_allListeners.Add(listener);
+
+            s_activeSources.EnumWithAction(source => {
+                if (listener.ListenToSource(source))
+                {
+                    source.AddListener(listener);
+                }
+            });
+
+            return listener;
+        }
+
+        internal void AddListener(ActivityListener listener)
+        {
+            if (_listeners == null)
+            {
+                Interlocked.CompareExchange(ref _listeners, new SynchronizedList<ActivityListener>(), null);
+            }
+
+            _listeners.AddIfNotExist(listener);
+        }
+
+        internal static void DetachListener(ActivityListener listener)
+        {
+            s_activeSources.EnumWithAction(source => {
+                var listeners = source._listeners;
+                listeners?.Remove(listener);
+            });
+            s_allListeners.Remove(listener);
```

Comment:
Removing from s_allListeners should happen first to prevent this ordering:
Thread 1 - listener is removed from all existing sources
Thread 2 - new source is created, listener is added to that source (never gets removed)
Thread 1 - listener is removed from s_allListeners list

Reply:
will do it.